pythonFiles/unittestadapter/utils.py

```diff
@@ -0,0 +1,193 @@
+            return str(lineno + i)
+    return "*"
+
+
+# Helper functions for test tree building.
+
+
+def build_test_node(path: str, name: str, type_: TestNodeTypeEnum) -> TestNode:
+    """Build a test node with no children. A test node can be a folder, a file or a class."""
+    return {
+        "path": path,
+        "name": name,
+        "type_": type_,
+        "children": [],
+    }
+
+
+def get_child_node(
+    name: str, path: str, type_: TestNodeTypeEnum, root: TestNode
+) -> TestNode:
+    """Find a child node in a test tree given its name and type. If the node doesn't exist, create it."""
+    try:
+        result = next(
+            node
+            for node in root["children"]
+            if node["name"] == name and node["type_"] == type_
+        )
+    except StopIteration:
+        result = build_test_node(path, name, type_)
+        root["children"].append(result)
+
+    return result  # type:ignore
+
+
+def build_test_tree(
+    suite: unittest.TestSuite, test_directory: str
+) -> Tuple[Union[TestNode, None], List[str]]:
+    """Build a test tree from a unittest test suite.
+
+    This function returns the test tree, and any errors found by unittest.
+    If no tests were discovered, return `None` and a list of errors (if any).
+
+    Test tree structure:
+    {
+        "path": <test_directory path>,
+        "type": "folder",
+        "name": <folder name>,
+        "children": [
+            { files and folders }
+            ...
+            {
+                "path": <file path>,
+                "name": filename.py,
+                "type_": "file",
+                "children": [
+                    {
+                        "path": <class path>,
+                        "name": <class name>,
+                        "type_": "class",
+                        "children": [
+                            {
+                                "path": <test path>,
+                                "name": <test name>,
+                                "type_": "test",
+                                "lineno": <line number>
+                                "id": <test case id>,
+                            }
+                        ]
+                    }
+                ]
+            }
+        ]
+    }
+    """
+    errors = []
+    directory_path = pathlib.PurePath(test_directory)
+    root = build_test_node(test_directory, directory_path.name, TestNodeTypeEnum.folder)
+
+    for test_case in get_test_case(suite):
+        test_id = test_case.id()
+        if test_id.startswith("unittest.loader._FailedTest"):
+            errors.append(str(test_case._exception))  # type: ignore
+        else:
+            # Get the static test path components: filename, class name and function name.
+            components = test_id.split(".")
+            filename, class_name, function_name = components[:3]
+            py_filename = f"{filename}.py"
+            folders = components[:-3]
```

Comment:
`components[:3]` gives you the filename, class name, and function name. But `components[:-3]` can  overlap with `[:3]` and yet is supposed to be only folders? Am I missing something or should it be `components[-3:]` for getting the the file name, etc.?

Reply:
> But components[:-3] can overlap with [:3] and yet is supposed to be only folders?

How can it overlap?

`folders` should only be the folder structure, so that when I rebuild the path to the test node I can concatenate the test directory + the content of `folders` + `py_filename`

Reply:
Here's an example of how your slicing could lead to overlap:

```Python
>>> components = ["A", "B", "C", "D", "E"]
>>> components[:3]
['A', 'B', 'C']
>>> components[:-3]
['A', 'B']
```

Notice how `A` and `B` end up in both slices.

Reply:
If the expectation is a list of folder, followed by the file name, class name, and function name, then this actually takes care of it nicely thanks to iterable unpacking:

```suggestion
            *folders, filename, class_name, function_name = components
            py_filename = f"{filename}.py"
```